### ISSUE TITLE
Estimation of the uncertainty in the predictions

### ISSUE THREAD

**Issue by schinto** [awslabs/autogluon#569]
Currently, Autogluon Tabular predicts values without giving an estimate of the uncertainty.
Are you going to integrate methods like [conformal prediction](https://arxiv.org/pdf/2005.07972.pdf) or another method to estimate the uncertainty in the prediction of new samples?

**Comment by Innixma** [comment]
This would be a very nice feature to have.

Before implementing more sophisticated methods of uncertainty, we can actually implement a simple method assuming the model is a bagged model: get the prediction of each fold model individually, then calculate the variance. This should give a rough approximation of model uncertainty.

To implement this, we just need to add a new argument to `BaggedEnsembleModel.predict_proba()` called `with_variance` to additionally return a variance calculation of the bagged predictions.

Contributions are welcome, as well as any insights into the best ways to calculate uncertainty!

**Comment by schinto** [comment]
Uncertainty estimation got available in catboost
https://towardsdatascience.com/tutorial-uncertainty-estimation-with-catboost-255805ff217e

https://youtu.be/aeMQk9EYqlw

**Comment by jwmueller** [comment]
Yes this catboost virtual-ensembles can be useful, particularly for quantifying model uncertainty which quantile regression does not measure.  But we should first add uncertainty-quantification as a bonus functionality for regular classification/regression, before trying to add it for quantile regression.  It'll presumably be a valuable functionality for all types of prediction problems.

**Comment by valeman** [comment]
https://github.com/valeman/awesome-conformal-prediction

Conformal Prediction is #1 uncertainty quantification method in 2021, has been implemented in several, libraries including MAPIE.

**Comment by jwmueller** [comment]
@valeman conformal calibration is now available in autogluon by setting `calibrate = True` (when used for quantile-regression problems), see our example: https://github.com/awslabs/autogluon/blob/master/examples/tabular/example_quantile_regression.py 

Implementation is thanks to @taesup-aws and based on paper:
Conformalized Quantile Regression (https://arxiv.org/abs/1905.03222)

**Comment by valeman** [comment]
That’s great thank you @jwmueller for sharing link to the example. Kudos to @taesup-aws i will including conformal powered auto gluon to software links for Amazing Conformal Prediction

**Comment by gomlfx** [comment]
I ran your code you supplied at the link below in colab and it gives
"Uknown keyword argument specificed: calibrate"

https://github.com/awslabs/autogluon/blob/master/examples/tabular/example_quantile_regression.py

ValueError: Unknown keyword argument specified: calibrate Valid kwargs:
['ag_args', 'ag_args_ensemble', 'ag_args_fit', 'auto_stack',
'excluded_model_types', 'feature_generator', 'holdout_frac',
'hyperparameter_tune_kwargs', 'keep_only_best', 'num_bag_folds',
'num_bag_sets', 'num_stack_levels', 'quantile_levels', 'refit_full',
'save_space', 'set_best_to_refit_full', 'unlabeled_data',
'use_bag_holdout', 'verbosity']

I use !pip install autogluon for colab, did I miss a different library I
need to install?

**Comment by valeman** [comment]
I wasn’t been able to install auto gluon on Colab today as well unfortunately. Is there clear instruction how install on Colab please?

**Comment by gomlfx** [comment]
Yes just

!pip install autogluon

**Comment by jwmueller** [comment]
@gomlfx This functionality is not available in the stable release of autogluon yet (will be in the next stable release). To use it now, you need to install either pip pre-release or from source following instructions here:  
https://auto.gluon.ai/dev/install.html

**Comment by jesuinovieira** [comment]
In addition to what has been done for quantile regression, are there any updates on this?

**Comment by alberto-jj** [comment]
Dear @jwmueller and @Innixma, could you kindly let me know if there is any update regarding the availability of conformal prediction or other uncertainty estimates for multi-class classification in AutoGluon? Perhaps for the upcoming versions? thanks for the efforts in your amazing library!

**Comment by jmakov** [comment]
Perhaps one could just wrap the models around https://github.com/scikit-learn-contrib/MAPIE ?

**Comment by Innixma** [comment]
MAPIE sounds interesting, might take a look at it.

**Comment by aabragan** [comment]
+1 I am extremely interested in a solution for this issue!

**Comment by solita-vilmaja** [comment]
+1 I also find this very interesting and useful!
